hack/airgap/main.go

```diff
@@ -0,0 +1,179 @@
+	// image signatures being verified.
+	noSignaturePolicy = `{"default":[{"type":"insecureAcceptAnything"}]}`
+
+	// requiredImages defines the list of images that must be saved
+	// into the output tar.
+	requiredImages = map[string]struct{}{
+		"docker.io/rancher/mirrored-pause":           struct{}{},
+		"docker.io/rancher/mirrored-coredns-coredns": struct{}{},
+	}
+
+	k3sVersion = regexp.MustCompile(`^v\d+\.\d+\.\d+\+k3s\d+$`)
+
+	fetcher = fetch
+)
+
+func main() {
+	if reexec.Init() {
+		return
+	}
+
+	version, output := "", ""
+
+	f := flag.NewFlagSet("", flag.ExitOnError)
+	f.StringVar(&version, "k3s-version", "", "k3s version to be used to pull the images file")
+	f.StringVar(&output, "output", "", "target file to be generated")
+	err := f.Parse(os.Args[1:])
+	if err != nil {
+		fmt.Println("failed to parse args %q: %v", os.Args, err)
+		os.Exit(1)
+	}
+
+	err = Save(version, output)
+	if err != nil {
+		fmt.Println("failed to save tarball: %v", err)
+		os.Exit(1)
+	}
+}
+
+// Save exports specific k3s container images into a tarball.
+// Only the images defined in requiredImages are exported.
+//
+// This is needed for Rancher Manager's airgap support and
+// replaces docker/podman's save command.
+// Note that both docker and podman do not support the execution
+// of save during a container image building process.
+func Save(version, output string) error {
+	if reexec.Init() {
+		return nil
+	}
+
+	path, err := setupStorage()
+	if err != nil {
+		return err
+	}
+
+	store, err := storage.GetStore(storage.StoreOptions{
+		GraphRoot: path,
+	})
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(path)
+
+	runtime, err := libimage.RuntimeFromStore(store, &libimage.RuntimeOptions{
+		SystemContext: &types.SystemContext{
+			SignaturePolicyPath: filepath.Join(path, "policy.json"),
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	imgs, err := k3sImages(version)
+	if err != nil {
+		return err
+	}
+
+	for _, img := range imgs {
+		fmt.Printf("pulling image %s\n", img)
+		_, err = runtime.Pull(context.TODO(), img, config.PullPolicyMissing, nil)
+		if err != nil {
+			return err
+		}
+	}
+
+	return runtime.Save(context.TODO(), imgs, "docker-archive", output, nil)
+}
+
+func fetch(url string) (io.ReadCloser, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), requestTimeout)
+	defer cancel()
+
+	fmt.Println("fetching", url)
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create request: %w", err)
+	}
+
+	res, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("request unsuccessful: %w", err)
+	}
+
+	return res.Body, nil
+}
+
+func k3sImages(version string) ([]string, error) {
+	if !k3sVersion.MatchString(version) {
+		return nil, fmt.Errorf("invalid k3s version: %s", version)
+	}
+
+	url := fmt.Sprintf("https://github.com/rancher/k3s/releases/download/%s/k3s-images.txt", version)
```

Comment:
Nice catch. The focus of this PR is to make the Dockerfiles self-contained with the minimum amount of changes. A clean-up PR is needed which can also cover this point, also updating (or removing) [the other reference](https://github.com/rancher/rancher/blob/1c25ef4d54a78aac88cc43abb097164f460b1076/.github/actions/k3s-images/action.yml#L12) to this endpoint.

Reply:
I believe one of the reasons to have it as `rancher/k3s` is that we apparently used to have a deal with docker so that repos on the `rancher` namespace wouldn't be affected by rate limiting, I think this might not be working anymore, since we've seen some rate limiting errors on our CI
https://www.suse.com/support/kb/doc/?id=000020630